### ISSUE TITLE
Periodic bursts of missing values that should be reported by the SNMP plugin

### ISSUE THREAD

**Issue by JamesFAtl** [collectd/collectd#2406]
*   Version of collectd: collectd-core 5.5.2
*   Operating system / distribution: Ubuntu 14.0.4

## Expected behavior

Consistent forwarding of sampled values to Graphite every 60 seconds.

## Actual behavior

Periodic bursts of missing values. Burst length varies from 1 to 15, usually in the range 1-5. Bursts are typically repeated every few minutes, over a period of a few hours.

These missing values only appear after a Collectd instance has been running for at least an hour, and only affect values sampled via SNMP. Each Collectd instance only samples a couple of dozen targets, so it's unlikely that we're overwhelming the core application.

Collectd is also _not_ reporting that entire collection runs are exceeding the configured interval, so we know it's not that kind of timeout.  We're not seeing synch failures logged for these targets - I know this, because we _do_ see these for targets that are unreachable.

There's also no correlation between metrics sampled via specific OIDs (e.g. PFE CPU utlilisation) or OIDs representing a row in ifTable or ifXTable.

What _is_ consistent is that if a subset of metrics for a row in a table are affected, it's consistently _that_ subset that's affected for the entire series of bursts.

## Steps to reproduce

TBD. I'm still establishing the boundaries of exactly where and when it happens, but it's independent of the make/model of the target device. We've also observed no consistency around what target device or metric is affected.

On the assumption that it's a scaling issue, I've cut down the configs on a staging instance to query just one target, and will ramp that up progressively in the hope of finding the threshold at which the problem occurs.

## Questions to help me isolate and/or fix the root problem

Is there a known scaling limit to the SNMP plugin, or a way to tune its buffers/caches? Because each device reports on dozens of interfaces, for both RX and TX stats, for 6-8 metrics, it's processing over a thousand metrics per device, which are then filtered via the pre-cache chain to ignore data for interfaces we don't care about.

Alternatively, is there a way of configuring it to report on things going awry? E.g, timeouts when sampling a particular OID from a table (say, IF-MIB::ifHCInOctets), or when a full set of values has been received but the plugin fails to process them.

**Comment by rubenk** [comment]
5.5 is quite old and we don't really support it anymore. Can you reproduce this with the latest version of collectd? We have packages for Ubuntu at https://github.com/collectd/collectd-ci

**Comment by JamesFAtl** [comment]
Thanks; I've upgraded to 5.7.2.20.g96b920e-1~trusty, and will monitor that over the next 24-48 hours.

I did get a stream of `format_graphite: error with uc_get_rate` errors in the logs, and no output of metrics. Changing `StoreRates` to False in the `write_graphite` configs fixed the problem, but it does suggest that #2209 isn't quite as fixed as one might hope. Should I raise a separate issue to track that, linked back to the existing one?

**Comment by JamesFAtl** [comment]
We're definitely still seeing missing metrics, even on that CI build.
I've cut its load back down to a single target host, and will watch for further evidence of a correlation between that and missing metrics.

**Comment by rubenk** [comment]
> Should I raise a separate issue to track that, linked back to the existing one?

Yes, please do.

**Comment by JamesFAtl** [comment]
Sorry, I've been away on leave for the last few days. I've just raised #2413 for the `error with uc_get_rate` issue, and we're still working to find the threshold at which values start to go missing. It's slow work because we're leaving 24 hours after each config change, to be sure we're not seeing a false negative.

**Comment by JamesFAtl** [comment]
The good news: collectd is in the clear.

We're still totally in the dark as to how we got the set of symptoms we had, but we fixed the problem with a combination of upgrading collectd to that CI build, and giving Graphite a set of hardware with much better disk I/O capabilities. We still get occasional gaps, but the rate of occurrence is close enough to the noise floor that it's no longer a practical problem for us.
